File: compiler/nodes/patterns/constructor.go
package patterns

import (
	"fmt"
	"reflect"

	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type ExtraElementFact struct{}

func (fact ExtraElementFact) String() string {
	return "is extra element"
}

type ConstructorPatternNode struct {
	ConstructorName string
	Elements        []database.Node
	Facts           *database.Facts

	matching database.Node

	matchingMarkerConstructor  *visit.MarkerConstructorDefinition
	matchingVariantConstructor *matchingVariantConstructor
}

type matchingVariantConstructor struct {
	index    int
	elements []matchingVariantConstructorElement
}

type matchingVariantConstructorElement struct {
	temporary database.Node
	pattern   database.Node
}

func (node *ConstructorPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseParameterizedConstructorPattern(parser *syntax.Parser) (*ConstructorPatternNode, *syntax.Error) {
	span := parser.Spanned()

	constructor, err := syntax.ParseConstructorName(parser)
	if err != nil {
		return nil, err
	}

	many, ok, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) ([]syntax.Many[database.Node, struct{}], *syntax.Error) {
		return syntax.ParseMany(parser, 1, ParseAtomicPattern, syntax.ParseNothing)
	})
	if err != nil {
		return nil, err
	}

	var elements []database.Node
	if ok {
		elements = make([]database.Node, 0, len(many))
		for _, item := range many {
			elements = append(elements, item.Value)
		}
	}

	return &ConstructorPatternNode{
		ConstructorName: constructor,
		Elements:        elements,
		Facts:           database.NewFacts(span()),
	}, nil
}

func ParseConstructorPattern(parser *syntax.Parser) (*ConstructorPatternNode, *syntax.Error) {
	span := parser.Spanned()

	constructor, err := syntax.ParseConstructorName(parser)
	if err != nil {
		return nil, err
	}

	return &ConstructorPatternNode{
		ConstructorName: constructor,
		Elements:        nil,
		Facts:           database.NewFacts(span()),
	}, nil
}

func (node *ConstructorPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	definition, ok := visit.ResolveOf(visitor, node.ConstructorName, node, []reflect.Type{
		reflect.TypeFor[*visit.MarkerConstructorDefinition](),
		reflect.TypeFor[*visit.VariantConstructorDefinition](),
	})

	if !ok {
		return
	}

	switch definition := definition.(type) {
	case *visit.MarkerConstructorDefinition:
		// No need to add a condition; markers only have one value

		for _, element := range node.Elements {
			visitor.VisitMatching(element)
			database.SetFact(element, ExtraElementFact{})
		}

		visitor.Constraint(typecheck.InstantiateConstraint(typecheck.Instantiation{
			Source:        node,
			Definition:    definition.Node,
			Substitutions: &map[database.Node]typecheck.Type{},
			Replacements:  map[database.Node]database.Node{definition.Node: node},
		}, visit.GetDefinitionConstraints))

		node.matchingMarkerConstructor = definition
	case *visit.VariantConstructorDefinition:
		matchingVariantConstructorElements := make([]matchingVariantConstructorElement, 0, len(node.Elements))
		for _, element := range node.Elements {
			temporary := visitor.VisitMatching(element)
			matchingVariantConstructorElements = append(matchingVariantConstructorElements, matchingVariantConstructorElement{
				temporary: temporary,
				pattern:   element,
			})
		}

		node.matchingVariantConstructor = &matchingVariantConstructor{
			index:    definition.Index,
			elements: matchingVariantConstructorElements,
		}

		if len(node.Elements) == 0 {
			visitor.Constraint(typecheck.InstantiateConstraint(typecheck.Instantiation{
				Source:        node,
				Definition:    definition.Node,
				Substitutions: &map[database.Node]typecheck.Type{},
				Replacements:  map[database.Node]database.Node{definition.Node: node},
			}, visit.GetDefinitionConstraints))
		} else {
			constructorNode := &database.HiddenNode{
				Facts: database.NewFacts(database.GetSpanFact(node)),
			}
			visitor.Db.Register(constructorNode)
			database.SetFact(constructorNode, typecheck.TypedFact{})

			visitor.Constraint(typecheck.InstantiateConstraint(typecheck.Instantiation{
				Source:        node,
				Definition:    definition.Node,
				Substitutions: &map[database.Node]typecheck.Type{},
				Replacements:  map[database.Node]database.Node{definition.Node: constructorNode},
			}, visit.GetDefinitionConstraints))

			visitor.Constraint(typecheck.TypeConstraint(constructorNode, typecheck.FunctionType[database.Node](node.Elements, node)))
		}
	}
}

func (node *ConstructorPatternNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	switch {
	case node.matchingMarkerConstructor != nil:
		// No code needed
		return nil
	case node.matchingVariantConstructor != nil:
		c.WriteString(span, " && (")
		c.WriteNode(span, node.matching)
		c.WriteString(span, "[runtime.variant] === ")
		c.WriteString(span, fmt.Sprintf("%d", node.matchingVariantConstructor.index))
		c.WriteString(span, ")")

		for index, element := range node.matchingVariantConstructor.elements {
			c.WriteString(span, " && ((")
			c.WriteNode(span, element.temporary)
			c.WriteString(span, " = ")
			c.WriteNode(span, node.matching)
			c.WriteString(span, "[")
			c.WriteString(span, fmt.Sprintf("%d", index))
			c.WriteString(span, "]) || true)")

			if err := c.Write(element.pattern); err != nil {
				return err
			}
		}

		return nil
	default:
		return c.Error(node)
	}
}

func (node *ConstructorPatternNode) EachTemporary(f func(database.Node)) {
	if node.matchingVariantConstructor != nil {
		for _, element := range node.matchingVariantConstructor.elements {
			f(element.temporary)
			EachTemporary(element.pattern, f)
		}
	}
}

File: compiler/src/nodes/patterns/constructor.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Db, Fact, HiddenNode, Node, NodeRef, Render},
    nodes::{Matching, Temporaries, parse_atomic_pattern, visit_pattern},
    syntax::{ParseError, Parser, parse_constructor_name},
    typecheck::{
        Instantation, InstantiateConstraint, Replacements, Substitutions, TypeConstraint, Typed,
    },
    visit::{
        Definition, MarkerConstructorDefinition, MatchPathSegment, VariantConstructorDefinition,
        Visit, Visitor,
    },
};

#[derive(Debug, Clone)]
pub struct ExtraElement;

impl Fact for ExtraElement {}

impl Render for ExtraElement {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "is extra element")
    }
}

#[derive(Debug, Clone)]
pub enum ConstructorMatch {
    Marker,
    Variant {
        index: usize,
        element_temporaries: Vec<NodeRef>,
    },
}

impl Fact for ConstructorMatch {}

impl Render for ConstructorMatch {}

#[derive(Debug)]
pub struct ConstructorPatternNode {
    pub constructor: String,
    pub elements: Vec<NodeRef>,
}

impl Node for ConstructorPatternNode {}

pub fn parse_parameterized_constructor_pattern(
    parser: &mut Parser<'_>,
) -> Result<ConstructorPatternNode, ParseError> {
    let constructor = parse_constructor_name(parser)?;

    let elements = parser
        .parse_optional(|parser| {
            parser.parse_many(1, parse_atomic_pattern, |parser| parser.parse_nothing())
        })?
        .unwrap_or_default()
        .into_iter()
        .map(|(node, _)| node)
        .collect::<Vec<_>>();

    Ok(ConstructorPatternNode {
        constructor,
        elements,
    })
}

pub fn parse_constructor_pattern(
    parser: &mut Parser<'_>,
) -> Result<ConstructorPatternNode, ParseError> {
    let constructor = parse_constructor_name(parser)?;

    Ok(ConstructorPatternNode {
        constructor,
        elements: Vec::new(),
    })
}

#[derive(Debug, Clone)]
enum ConstructorDefinition {
    Marker(MarkerConstructorDefinition),
    Variant(VariantConstructorDefinition),
}

impl Visit for ConstructorPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        let Some(definition) =
            visitor.resolve(&self.constructor, node, |definition| match definition {
                Definition::MarkerConstructor(definition) => {
                    Some(ConstructorDefinition::Marker(definition.clone()))
                }
                Definition::VariantConstructor(definition) => {
                    Some(ConstructorDefinition::Variant(definition.clone()))
                }
                _ => None,
            })
        else {
            visit_pattern(node, visitor, None);
            return;
        };

        match definition {
            ConstructorDefinition::Marker(definition) => {
                visit_pattern(node, visitor, Some(MatchPathSegment::Match));

                for element in &self.elements {
                    visitor.visit_matching(element, None);
                    visitor.edge(element, node, "element");
                    visitor.insert(element, ExtraElement);
                }

                visitor.constraint(InstantiateConstraint::new(Instantation {
                    source_node: node.clone(),
                    definition: definition.node.clone(),
                    substitutions: Substitutions::new(),
                    replacements: Replacements::from_iter([(definition.node, node.clone())]),
                    from_expression: true,
                }));

                visitor.insert(node, ConstructorMatch::Marker);
            }
            ConstructorDefinition::Variant(definition) => {
                visit_pattern(
                    node,
                    visitor,
                    self.elements
                        .is_empty()
                        .then(|| MatchPathSegment::Variant(definition.variant.clone())),
                );

                let element_temporaries = self
                    .elements
                    .iter()
                    .enumerate()
                    .map(|(index, element)| {
                        let temporary = visitor.visit_matching(
                            element,
                            Some(MatchPathSegment::VariantElement(
                                definition.variant.clone(),
                                index,
                                self.elements.len(),
                            )),
                        );

                        visitor.edge(element, node, "element");

                        temporary
                    })
                    .collect::<Vec<_>>();

                if self.elements.is_empty() {
                    visitor.constraint(InstantiateConstraint::new(Instantation {
                        source_node: node.clone(),
                        definition: definition.variant.clone(),
                        substitutions: Substitutions::new(),
                        replacements: Replacements::from_iter([(
                            definition.variant.clone(),
                            node.clone(),
                        )]),
                        from_expression: true,
                    }));
                } else {
                    let span = visitor.span(node);
                    let constructor_node = visitor.node(span, HiddenNode(None));
                    visitor.insert(&constructor_node, Typed::default());

                    visitor.constraint(InstantiateConstraint::new(Instantation {
                        source_node: node.clone(),
                        definition: definition.variant.clone(),
                        substitutions: Substitutions::new(),
                        replacements: Replacements::from_iter([(
                            definition.variant.clone(),
                            constructor_node.clone(),
                        )]),
                        from_expression: true,
                    }));

                    visitor.constraint(TypeConstraint::new(
                        constructor_node,
                        visitor.function_type(self.elements.iter().cloned(), node.clone()),
                    ));
                }

                visitor.insert(
                    node,
                    ConstructorMatch::Variant {
                        index: definition.index,
                        element_temporaries: element_temporaries.clone(),
                    },
                );

                visitor.insert(
                    node,
                    Temporaries {
                        has: element_temporaries,
                        inherit: self.elements.clone(),
                    },
                )
            }
        }
    }
}

impl Codegen for ConstructorPatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Matching(matching) = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        let constructor_match = ctx
            .get::<ConstructorMatch>(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        match constructor_match {
            ConstructorMatch::Marker => {}
            ConstructorMatch::Variant {
                index: variant,
                element_temporaries,
            } => {
                ctx.condition(ir::Condition::EqualToVariant {
                    input: matching.clone(),
                    variant,
                });

                for (index, (element, temporary)) in
                    self.elements.iter().zip(element_temporaries).enumerate()
                {
                    ctx.condition(ir::Condition::Initialize {
                        variable: temporary,
                        node: None,
                        value: ir::Value::VariantElement {
                            input: matching.clone(),
                            variant,
                            index,
                        },
                        mutable: false,
                    });

                    ctx.codegen(element)?;
                }
            }
        }

        Ok(())
    }
}
